### ISSUE TITLE
Rustc can't infer GAT type when used as argument to Fn closure outside of trait

### ISSUE THREAD

**Issue by geom3trik** [rust-lang/rust#88459]
I tried this code:

[playgorund](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0dcd779ee08f5b25260c4c212228933c)
```rust
#![feature(generic_associated_types)]

trait Trait {
    type Assoc<'a>;
}

fn f<T: Trait>(_: T, _: impl Fn(T::Assoc<'_>)) {}

struct Type;

impl Trait for Type {
    type Assoc<'a> = ();
}

fn main() {
    f(Type, |_|());
}
```

I expected to see this happen: compiles with no errors.

Instead, this happened: 
```
error[E0631]: type mismatch in closure arguments
  --> src/main.rs:16:5
   |
16 |     f(Type, |_|());
   |     ^       ----- found signature of `fn(()) -> _`
   |     |
   |     expected signature of `for<'r> fn(<Type as Trait>::Assoc<'r>) -> _`
   |
note: required by a bound in `f`
  --> src/main.rs:7:30
   |
7  | fn f<T: Trait>(_: T, _: impl Fn(T::Assoc<'_>)) {}
   |                              ^^^^^^^^^^^^^^^^ required by this bound in `f`
   ```
The error seems to suggest that rustc can't figure out that the associated type is `()` which should satisfy the bound in `f`.

### Meta
1.56.0-nightly ( https://github.com/rust-lang/rust/commit/5eacec9ec7e112a0de1011519a57c45586d58414 )

This could be related to #85921, but unlike that issue this one does not seem to be fixed on latest nightly.

**Comment by b-naber** [comment]
This seems related to https://github.com/rust-lang/rust/issues/88382
